### PR TITLE
Feature/quiz resume

### PR DESCRIPTION
**Pull request by eri-trabiccolo**
Quiz Resume feature

### CODE REVIEW

**Comment by eri-trabiccolo** [comment]
to test this run `npm run build`.

**Comment by eri-trabiccolo** [comment]
@actuallyakash is this ready to be reviewed or is it still a draft?

**Comment by actuallyakash** [comment]
> @actuallyakash is this ready to be reviewed or is it still a draft?

sorry, I missed it. Yes, it's ready for review.

**Review by eri-trabiccolo — COMMENTED** [review]
This is in good shape. It'd require some more changes done in #2562, but I haven't had the time to review that PR thoroughly.
@actuallyakash please stress test this and #2562.
Thanks for your work!

**Comment by actuallyakash** [comment]
Got it @eri-trabiccolo 

All thanks to you for all the support <3

File: assets/js/llms-quiz.js [hunk 1]
```diff
@@ -337,10 +355,46 @@
 				prev_id = ids[ curr - 1 ];
 			}
 
-			setTimeout( function() {
-				self.toggle_loader( 'hide' );
-				self.load_question( self.questions[ prev_id ] );
-			}, 100 );
+			// Retrieve previous question HTML from the server.
+			if ( ! self.questions[ prev_id ] ) {
```
Comment (brianhogg) [inline assets/js/llms-quiz.js]: How was `self.questions` populated before?

File: assets/js/builder/Models/Quiz.js [hunk 2]
```diff
@@ -52,6 +54,7 @@ define( [
 		 * @since 3.16.0
 		 * @since 7.4.0 Added filter for filtering defaults.
 		 * @since [version] Replaced unused `random_answers` property with `random_questions`.
+		 *              Added filter for filtering defaults and `can_be_resumed` property.
```
Comment (brianhogg) [inline assets/js/builder/Models/Quiz.js]: Undo this file. No actual changes.

File: includes/abstracts/llms.abstract.notification.view.quiz.completion.php [hunk 3]
```diff
@@ -133,7 +134,7 @@ protected function set_merge_data( $code ) {
 				break;
 
 			case '{{GRADE}}':
-				$code = round( $attempt->get( 'grade' ), 2 ) . '%';
+				$code = round( $attempt->get( 'grade' ) ?? 0, 2 ) . '%';
```
Comment (brianhogg) [inline includes/abstracts/llms.abstract.notification.view.quiz.completion.php:137]: Do we want to show this as 0% or blank? 0% implies they took it but got zero questions correct.

File: includes/admin/reporting/tables/llms.table.quiz.attempts.php [hunk 4]
```diff
@@ -117,8 +119,12 @@ protected function get_data( $key, $attempt ) {
 				break;
 
 			case 'grade':
-				$value  = $attempt->get( $key ) ? $attempt->get( $key ) . '%' : '0%';
-				$value .= ' (' . $attempt->l10n( 'status' ) . ')';
+				$value      = $attempt->get( $key ) ? $attempt->get( $key ) . '%' : '0%';
```
Comment (brianhogg) [inline includes/admin/reporting/tables/llms.table.quiz.attempts.php:122]: Again we probably don't want to show `0%` here if not actually finished the attempt.

File: includes/class.llms.ajax.handler.php [hunk 5]
```diff
@@ -669,15 +669,167 @@ public static function quiz_start( $request ) {
...
+			);
+
+			return array(
+				'attempt_key'    => $attempt->get_key(),
+				'html'           => $html,
+				'question_id'    => $question_id,
+				'total'          => $attempt->get_count( 'questions' ),
+				'question_ids'   => $question_ids,
+				'can_be_resumed' => $attempt->can_be_resumed(),
+			);
```
Comment (brianhogg) [inline includes/class.llms.ajax.handler.php]: Early return vs. if/else?

File: includes/class.llms.ajax.handler.php [hunk 6]
```diff
@@ -669,15 +669,167 @@ public static function quiz_start( $request ) {
...
+					'question' => llms_get_post( $question_id ),
+				)
+			);
+
+			return array(
+				'html'        => $html,
+				'question_id' => $question_id,
+			);
+
+		} else {
```
Comment (brianhogg) [inline includes/class.llms.ajax.handler.php]: Remove else

File: includes/class.llms.install.php [hunk 7]
```diff
@@ -448,6 +449,7 @@ private static function get_schema() {
   `status` varchar(15) DEFAULT '',
   `attempt` bigint(20) DEFAULT NULL,
   `grade` float DEFAULT NULL,
+  `can_be_resumed` tinyint(1) DEFAULT '0',
```
Comment (brianhogg) [inline includes/class.llms.install.php:446]: Check that this will automatically get picked up as a new column without a migration.

File: includes/class.llms.ajax.handler.php [hunk 8]
```diff
@@ -639,7 +639,7 @@ public static function quiz_start( $request ) {
 			$attempt = $student->quizzes()->get_attempt_by_key( $request['attempt_key'] );
 		}
 
-		if ( ! $attempt || 'new' !== $attempt->get_status() ) {
+		if ( ! $attempt || 'new' !== $attempt->get( 'status' ) ) {
```
Comment (brianhogg) [inline includes/class.llms.ajax.handler.php:631]: Why was this changed?

File: includes/class.llms.ajax.handler.php [hunk 9]
```diff
@@ -669,15 +669,167 @@ public static function quiz_start( $request ) {
...
+			$err->add( 400, __( 'Attempt key is required.', 'lifterlms' ) );
+			return $err;
+		}
+
+		$attempt = $student->quizzes()->get_attempt_by_key( $request['attempt_key'] );
+
+		if ( empty( $attempt ) ) {
+			$err->add( 404, __( 'The requested attempt could not be found.', 'lifterlms' ) );
+			return $err;
+		}
```
Comment (brianhogg) [inline includes/class.llms.ajax.handler.php:709]: We might want to be careful about giving too many specifics on what is/isn't found. The attempt key is only three characters long I believe and relatively easy to brute force.

File: includes/class.llms.quiz.data.php [hunk 10]
```diff
@@ -21,11 +21,28 @@
...
+	public $quiz_id;
+
+	/**
+	 * Quiz instance.
 	 *
-	 * @since    3.16.0
+	 * @since [version]
+	 * @var LLMS_Quiz
+	 */
+	public $quiz;
```
Comment (brianhogg) [inline includes/class.llms.quiz.data.php]: Do we want these as public properties without an accessor?

File: assets/js/builder/Views/Question.js [hunk 11]
```diff
@@ -126,7 +129,8 @@ define( [
 				this.init_editor( 'question-clarifications--' + this.model.get( 'id' ), {
 					mediaButtons: false,
 					tinymce: {
-						toolbar1: 'bold,italic,strikethrough,bullist,numlist,alignleft,aligncenter,alignright',
+						plugins: 'image',
+						toolbar1: 'bold,italic,strikethrough,bullist,numlist,alignleft,aligncenter,alignright,image',
```
Comment (brianhogg) [inline assets/js/builder/Views/Question.js]: This seems outside of the scope of this change. Why do we need an image upload in the tiny MCE editor for question? Might already be a thing.

File: includes/class.llms.query.quiz.attempt.php [hunk 12]
```diff
@@ -200,6 +203,11 @@ protected function sql_where() {
 			$sql     .= " AND status NOT IN ({$prepared})";
 		}
 
+		$can_be_resumed = $this->get( 'can_be_resumed' );
+		if ( '' !== $can_be_resumed ) {
```
Comment (brianhogg) [inline includes/class.llms.query.quiz.attempt.php:207]: Be more explicit here. `is_numeric` check since we're using `%d` in the prepare?

File: includes/models/model.llms.quiz.attempt.php [hunk 13]
```diff
@@ -747,7 +825,65 @@ public function is_passing() {
...
+
+		/**
+		 * Filters the X time for resuming quiz.
+		 *
+		 * @since [version]
+		 *
+		 * @param int $time_period The time period in days.
+		 */
+		$time_period     = apply_filters( 'llms_quiz_attempt_resume_time_period', 1, $this );
+		$expiration_date = strtotime( "+{$time_period} days", strtotime( $start_date ) );
```
Comment (brianhogg) [inline includes/models/model.llms.quiz.attempt.php]: Sanitize this value

File: includes/models/model.llms.quiz.php [hunk 14]
```diff
@@ -222,6 +296,8 @@ public function is_open( $user_id = null ) {
 		/**
 		 * Filters whether the quiz is open to a student or not.
 		 *
+		 * @since 3.37.2
```
Comment (brianhogg) [inline includes/models/model.llms.quiz.php]: Why was this added?

File: includes/models/model.llms.student.quizzes.php [hunk 15]
```diff
@@ -5,13 +5,13 @@
  * @package LifterLMS/Models/Classes
  *
  * @since 3.9.0
- * @version 6.4.0
+ * @version [version]
```
Comment (brianhogg) [inline includes/models/model.llms.student.quizzes.php]: Undo changes to this file. No actual changes made, just spacing.

File: templates/admin/reporting/tabs/quizzes/attempt.php [hunk 16]
```diff
@@ -154,6 +159,12 @@
 					</span>
 			</button>
 			<?php endif; ?>
+			<?php if ( $attempt->can_be_resumed() ) : // Show the clear resume attempt button only if quiz can be resumed. ?>
+			<button class="llms-button-secondary large" name="llms_quiz_attempt_action" type="submit" value="llms_disable_resume_attempt">
+				<i class="fa fa-ban" aria-hidden="true"></i>
+				<?php _e( 'Disable Resume Attempt', 'lifterlms' ); ?>
```
Comment (brianhogg) [inline templates/admin/reporting/tabs/quizzes/attempt.php]: Escape this.

File: templates/content-single-question.php [hunk 17]
```diff
@@ -18,7 +20,7 @@
  */
 do_action( 'lifterlms_single_question_before_summary', $args ); ?>
 
-	<h3 class="llms-question-text"><?php echo $question->get_question(); ?></h3>
+	<h3 class="llms-question-text"><?php echo $question->get_question( 'html', $attempt ); ?></h3>
```
Comment (brianhogg) [inline templates/content-single-question.php]: Rename to get_question_from_attempt instead rather than having two params?

File: templates/quiz/questions/content-choice.php [hunk 18]
```diff
@@ -4,24 +4,33 @@
...
-$input_type = ( 'yes' === $question->get( 'multi_choices' ) ) ? 'checkbox' : 'radio';
+$input_type      = ( 'yes' === $question->get( 'multi_choices' ) ) ? 'checkbox' : 'radio';
+$question_answer = $attempt ? $attempt->get_question_answer( $question->get( 'id' ) ) : [];
 ?>
 
 <ol class="llms-question-choices">
 	<?php foreach ( $question->get_choices() as $choice ) : ?>
-
+		<?php
+		$answer = is_array( $question_answer ) ? in_array( $choice->get( 'id' ), $question_answer, true ) ? $choice->get( 'id' ) : null : null;
```
Comment (brianhogg) [inline templates/quiz/questions/content-choice.php:24]: Might want to split this up for clarity, hard to grok.

File: templates/quiz/questions/content-picture_choice.php [hunk 19]
```diff
@@ -6,25 +6,34 @@
...
 
 defined( 'ABSPATH' ) || exit;
 
-$input_type = ( 'yes' === $question->get( 'multi_choices' ) ) ? 'checkbox' : 'radio';
-$choices    = $question->get_choices();
-$cols       = llms_get_picture_choice_question_cols( count( $choices ) );
+$input_type      = ( 'yes' === $question->get( 'multi_choices' ) ) ? 'checkbox' : 'radio';
+$choices         = $question->get_choices();
+$cols            = llms_get_picture_choice_question_cols( count( $choices ) );
+$question_answer = $attempt ? $attempt->get_question_answer( $question->get( 'id' ) ) : [];
```
Comment (brianhogg) [inline templates/quiz/questions/content-picture_choice.php]: For this and the other, how was the selected answer shown before? Client-side JS vs. server-side HTML?

File: templates/quiz/results-attempt.php [hunk 20]
```diff
@@ -18,20 +20,22 @@
...
-	<?php if ( $attempt->get_count( 'available_points' ) ) : ?>
-		<?php echo llms_get_donut( $attempt->get( 'grade' ), $attempt->l10n( 'status' ), 'default', array( $donut_class ) ); ?>
-	<?php endif; ?>
-	<ul class="llms-quiz-meta-info">
-		<?php if ( $attempt->get_count( 'gradeable_questions' ) ) : ?>
-			<li class="llms-quiz-meta-item"><?php printf( __( 'Correct Answers: %1$d / %2$d', 'lifterlms' ), $attempt->get_count( 'correct_answers' ), $attempt->get_count( 'gradeable_questions' ) ); ?></li>
+<?php if ( ! $attempt->can_be_resumed() ) { ?>
+	<aside class="llms-quiz-results-aside">
+		<?php if ( $attempt->get_count( 'available_points' ) ) : ?>
+			<?php echo llms_get_donut( $attempt->get( 'grade' ), $attempt->l10n( 'status' ), 'default', array( $donut_class ) ); ?>
```
Comment (brianhogg) [inline templates/quiz/results-attempt.php]: Escape

File: templates/quiz/results.php [hunk 21]
```diff
@@ -72,7 +73,7 @@
 				<?php foreach ( $attempts as $attempt ) : ?>
 					<option value="<?php echo esc_url( $attempt->get_permalink() ); ?>">
 						<?php // Translators: %1$d = Attempt number; %2$s = Grade percentage; %3$s = Pass/fail text. ?>
-						<?php printf( __( 'Attempt #%1$d - %2$s (%3$s)', 'lifterlms' ), $attempt->get( 'attempt' ), round( $attempt->get( 'grade' ), 2 ) . '%', $attempt->l10n( 'status' ) ); ?>
+						<?php printf( __( 'Attempt #%1$d - %2$s (%3$s)', 'lifterlms' ), $attempt->get( 'attempt' ), round( $attempt->get( 'grade' ) ?? 0, 2 ) . '%', $attempt->l10n( 'status' ) ); ?>
```
Comment (brianhogg) [inline templates/quiz/results.php]: Still show 0% when incomplete?

File: templates/quiz/start-button.php [hunk 22]
```diff
@@ -31,10 +33,10 @@
 
 	<?php if ( $quiz ) : ?>
 
-		<form method="POST" action="" name="llms_start_quiz" enctype="multipart/form-data">
-
-			<?php if ( $quiz->is_open() ) : ?>
+		<?php $start_button_string = $quiz->can_be_resumed_by_student() ? esc_html__( 'Restart Quiz', 'lifterlms' ) : esc_html__( 'Start Quiz', 'lifterlms' ); ?>
```
Comment (brianhogg) [inline templates/quiz/start-button.php]: Escape when output not here. It's being double escaped.

File: templates/quiz/start-button.php [hunk 23]
```diff
@@ -53,23 +55,48 @@
...
 
 	<?php else : ?>
 
-		<p><?php _e( 'You are not able to take this quiz', 'lifterlms' ); ?></p>
+		<p><?php esc_html_e( 'You are not able to take this quiz', 'lifterlms' ); ?></p>
+
+	<?php endif; ?>
 
+	<?php if ( $lesson->get_next_lesson() && llms_is_complete( get_current_user_id(), $lesson->get( 'id' ), 'lesson' ) ) : ?>
+		<a href="<?php echo get_permalink( $lesson->get_next_lesson() ); ?>" class="button llms-button-secondary llms-next-lesson"><?php esc_html_e( 'Next Lesson', 'lifterlms' ); ?></a>
```
Comment (brianhogg) [inline templates/quiz/start-button.php]: Escape.

File: assets/js/builder/Views/Question.js (hunk 11 above)
Comment (brianhogg) [reply assets/js/builder/Views/Question.js]: This was adding a new image plugin to TinyMCE rather than just having mediaButtons be `true`

File: templates/quiz/questions/content-picture_choice.php (hunk 19 above)
Comment (brianhogg) [reply templates/quiz/questions/content-picture_choice.php]: Via getting the HTML when clicking next question.

File: includes/class.llms.query.quiz.attempt.php (hunk 12 above)
Comment (brianhogg) [reply includes/class.llms.query.quiz.attempt.php:207]: It's the same format as other conditions like `attempt`
